src/controllers/mainController.ts

```diff
@@ -113,10 +114,52 @@ export default class MainController implements vscode.Disposable {
                 this._event.on(Constants.cmdShowGettingStarted, () => { self.launchGettingStartedPage(); });
                 this.registerCommand(Constants.cmdNewQuery);
-                this._event.on(Constants.cmdNewQuery, () => { self.runAndLogErrors(self.onNewQuery(), 'onNewQuery'); });
+                this._event.on(Constants.cmdNewQuery, () => {
+                    self.runAndLogErrors(self.onNewQuery(), 'onNewQuery');
+                });
                 this.registerCommand(Constants.cmdRebuildIntelliSenseCache);
                 this._event.on(Constants.cmdRebuildIntelliSenseCache, () => { self.onRebuildIntelliSense(); });
 
+                // register the object explorer tree provider
+                this._objectExplorerProvider = new ObjectExplorerProvider(this._connectionMgr);
+                this._context.subscriptions.push(
+                    vscode.window.registerTreeDataProvider('objectExplorer', this._objectExplorerProvider)
+                );
+                this.registerCommand(Constants.cmdAddObjectExplorer);
+                this._event.on(Constants.cmdAddObjectExplorer, async () => {
+                    if (self._objectExplorerProvider.objectExplorerExists) {
+                        return self._objectExplorerProvider.createSession();
+                    } else {
+                        self._objectExplorerProvider.objectExplorerExists = true;
+                        return vscode.commands.executeCommand(Constants.cmdOpenObjectExplorerCommand);
+                    }
+                });
+                this.registerCommand(Constants.cmdObjectExplorerNewQuery);
+                this._event.on(Constants.cmdObjectExplorerNewQuery, () => {
+                    // emit new query with sessionId for connection
+                    self.runAndLogErrors(self.onNewQuery(this._objectExplorerProvider.currentNode.sessionId), 'onNewQuery');
+                });
+                this.registerCommand(Constants.cmdRemoveObjectExplorerNode);
+                this._event.on(Constants.cmdRemoveObjectExplorerNode, () => {
+                    return this._objectExplorerProvider.removeObjectExplorerNode(this._objectExplorerProvider.currentNode);
+                });
+                this.registerCommand(Constants.cmdRefreshObjectExplorerNode);
+                this._event.on(Constants.cmdRefreshObjectExplorerNode, () => {
+                    return this._objectExplorerProvider.refreshNode(this._objectExplorerProvider.currentNode);
+                });
+                this._context.subscriptions.push(vscode.commands.registerCommand(Constants.cmdScriptSelect, async (treeNodeInfo) => {
+                    const objectNames = treeNodeInfo.label.split('.');
+                    const tableName = `[${escapeCharacters(objectNames[0])}].[${escapeCharacters(objectNames[1])}]`;
+                    const databaseName =  treeNodeInfo.parentNode.parentNode.nodeType === Constants.databaseString ?
```

Comment:
It might be a good idea to 

1. Check that the current treeNodeInfo.nodeType === Constants.table
2. Instead of assuming that the parent DB is exactly two nodes above, just walk the parents until you find one with nodeType === Constants.database

And then probably display an error if either of those fail (mostly so users are more likely to report the issue and we can investigate)

It's probably not strictly necessary but I really prefer we be safer on things like this since if this did change and start failing for some reason it'd be more annoying to track down without a clear error message.